0.data_analysis_and_processing/nbconverted/filter_nuclei.py

```diff
@@ -0,0 +1,183 @@
+
+
+# In[5]:
+
+
+def filter_bounding_box_size(_scdf, _bounding_box_col):
+    """
+    Filter nuclei if the robust z score of a bounding box dimension is less than three.
+
+    Parameters
+    ----------
+    _scdf: Pandas Dataframe
+        The unfiltered single-cell data containing bounding box dimensions.
+
+    _bounding_box_col: String
+        Column specifying the bounding box dimension.
+
+    Returns
+    -------
+    Filtered single-cell pandas dataframe.
+    """
+
+    median = scdfs[_bounding_box_col].median()
+
+    # Calculate the absolute deviations from the median
+    _scdf[f"{_bounding_box_col}_absolute_deviation"] = np.abs(_scdf[_bounding_box_col] - median)
+
+    # Calculate the MAD
+    mad = _scdf[f"{_bounding_box_col}_absolute_deviation"].median()
+
+    # Calculate the robust Z-score
+    _scdf[f"{_bounding_box_col}_robust_z_score"] = (_scdf[_bounding_box_col] - median) / mad
+
+    # Returns the single cells with a robust z score less than three
+    return _scdf.loc[_scdf[f"{_bounding_box_col}_robust_z_score"] < 3]
+
+
+# ## Retain only Common Metadata and Bounding Box Columns
+
+# In[6]:
+
+
+bounding_box_cols = ["Nuclei_AreaShape_BoundingBoxMaximum_X", "Nuclei_AreaShape_BoundingBoxMinimum_X", "Nuclei_AreaShape_BoundingBoxMaximum_Y", "Nuclei_AreaShape_BoundingBoxMinimum_Y"]
+
+common_cols = set(scdfs[0].columns[scdfs[0].columns.str.contains('Metadata')])
+
+for scdf in scdfs[1:]:
+    common_cols = set(scdf.columns[scdf.columns.str.contains('Metadata')]) & set(scdf.columns)
+
+scdfs = pd.concat(scdfs, axis=0)[list(common_cols) + bounding_box_cols]
+
+
+# ## Calculate Bounding Box Dimensions
+
+# In[7]:
+
+
+scdfs["Nuclei_AreaShape_BoundingBoxDelta_X"] = scdfs["Nuclei_AreaShape_BoundingBoxMaximum_X"] - scdfs["Nuclei_AreaShape_BoundingBoxMinimum_X"]
+scdfs["Nuclei_AreaShape_BoundingBoxDelta_Y"] = scdfs["Nuclei_AreaShape_BoundingBoxMaximum_Y"] - scdfs["Nuclei_AreaShape_BoundingBoxMinimum_Y"]
+
+
+# In[8]:
+
+
+scdfs.head()
+
+
+# # Bounding Box Dimension Distributions
+# Visualizes bounding box distributions before and after filtering and prints out cell counts information.
+
+# In[9]:
+
+
+pre_filter_sc_count = scdfs.shape[0]
+
+g = sns.jointplot(data=scdfs, x="Nuclei_AreaShape_BoundingBoxDelta_X", y="Nuclei_AreaShape_BoundingBoxDelta_Y")
+g.fig.set_size_inches(18, 10)
+g.set_axis_labels("Bounding Box Height (Pixels)", "Bounding Box Length (Pixels)")
+g.fig.suptitle("Distribution of Bounding Box Sizes before Filtering")
+plt.savefig(filtered_sc_figure_path / "nuclei_dimension_distributions_before_filtering.png")
+plt.show()
+
+
+# ## Filter Nuclei by Height and Length
+
+# In[10]:
+
+
+scdfs = filter_bounding_box_size(_scdf=scdfs, _bounding_box_col="Nuclei_AreaShape_BoundingBoxDelta_X")
+scdfs = filter_bounding_box_size(_scdf=scdfs, _bounding_box_col="Nuclei_AreaShape_BoundingBoxDelta_Y")
+
+scdfs.filter(like="Metadata").to_parquet(filtered_sc_path / "filtered_single_cell_profiles.parquet")
```

Comment:
Recommend including `index=False` and resaving this data so there isn't a random "index" column in the file.

Reply:
Will change this